framework/include/utils/RankMap.h

```diff
@@ -0,0 +1,52 @@
+#ifndef RANKMAP_H
+#define RANKMAP_H
+
+#include "PerfGraphInterface.h"
+
+#include "libmesh/parallel_object.h"
+
+/**
+ * Builds lists and maps that help in knowing which physical hardware nodes each rank is on.
+ *
+ * Note: large chunks of this code were originally committed by @dschwen in PR #12351
+ *
+ * https://github.com/idaholab/moose/pull/12351
+ */
+class RankMap : ParallelObject, PerfGraphInterface
+{
+public:
+  /**
+   * Constructs and fills the map
+   */
+  RankMap(const Parallel::Communicator & comm, PerfGraph & perf_graph);
+
+  /**
+   * Returns the "hardware ID" (a unique ID given to each physical compute node in the job)
+   * for a given processor ID (rank)
+   */
+  unsigned int hardwareID(processor_id_type pid) const { return _rank_to_hardware_id[pid]; }
+
+  /**
+   * Returns the ranks that are on the given hardwareID (phsical node in the job)
+   */
+  const std::vector<processor_id_type> & ranks(unsigned int hardware_id) const
+  {
+    return _hardware_id_to_ranks.at(hardware_id);
```

Comment:
Are you sure you want to use `at()`? It throws and doesn't terminate MPI "well" in parallel. If you intend for this to be a MOOSE utility it probably should be changed to provide a better error message.

Reply:
I don't think this will be widely used outside of the MOOSE team so it's fine if you want to leave it.

Reply:
I'm ok with it throwing... but you can tell me if you'd rather have me change it.

It has to be `.at()` or some `find()` thing though because the function is const.

Reply:
Right, we've switched a few of these to this:

```
auto item = map.find()
if (item == map.end())
   mooseError("Item not found");
return *item; // or item->first
```

If you ever hit one of these `at()` in parallel it can be infuriating. If a non-zero ranks gets there first, your application will just quit and you won't get any errors in the error log, nothing...

Reply:
Sure - I'll change it 👍